### PR TITLE
feat/enterpriseportal: implement SubscriptionLicenseChecksService

### PR DESCRIPTION
**Pull request by bobheadxi**
Implements the RPC defined in https://github.com/sourcegraph/sourcegraph/pull/64396. Follow-up PRs will implement migrations for in-instance checks and for through-dotcom checks.

One major change is that we now bypass the check for subscriptions that are denoted as associated with `INTERNAL` instances.

Most of the diff is generated mocks.

Part of https://linear.app/sourcegraph/issue/CORE-227

## Test plan

- [x] Unit tests
- [x] E2E tests (`sg test enterprise-portal-e2e`)

image

### CODE REVIEW

**Comment by bobheadxi** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64400?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#64400** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64400?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#64396** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64396?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64253** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64253?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#64057](https://github.com/sourcegraph/sourcegraph/pull/64057) <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64057?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#64115](https://github.com/sourcegraph/sourcegraph/pull/64115) <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64115?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#63959** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63959?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64114** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64114?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#63927** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63927?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64087** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64087?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64317** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64317?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64090** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64090?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#64376** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64376?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bobheadxi and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by eseliger — APPROVED** [review]

File: cmd/enterprise-portal/internal/database/subscriptions/licenses.go [hunk 1]
```diff
@@ -442,3 +469,61 @@ WHERE id = @licenseID
...
+	Time *utctime.Time
+}
+
+// SetDetectedInstance sets the instance ID that was detected to be using this
+// license.
+func (s *LicensesStore) SetDetectedInstance(ctx context.Context, licenseID string, opts SetDetectedInstanceOpts) error {
+	if opts.Time == nil {
+		opts.Time = pointers.Ptr(utctime.Now())
+	} else if opts.Time.GetTime().After(time.Now()) {
+		return errors.New("revocation time cannot be in the future")
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply cmd/enterprise-portal/internal/database/subscriptions/licenses.go]: This is some copy-pasta from elsewhere, but removed the check for now! I think generally the time is left as nil

File: cmd/enterprise-portal/e2e/e2e_test.go [hunk 2]
```diff
@@ -80,9 +84,15 @@ func newE2EClients(t *testing.T) *Clients {
 	require.NoError(t, err)
 	t.Cleanup(func() { _ = client.Close() })
 
+	clientWithoutCreds, err := grpc.NewClient("dns:///"+addr.Host,
```
Comment (unknwon) [inline cmd/enterprise-portal/e2e/e2e_test.go:87]: TIL `dns://` works here...

File: cmd/enterprise-portal/internal/subscriptionlicensechecksservice/v1_store.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		LicenseKey:  licenseKey,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(keys) == 0 {
+		return nil, nil
+	}
+	if len(keys) > 1 {
+		spew.Dump(keys)
```
Comment (unknwon) [inline cmd/enterprise-portal/internal/subscriptionlicensechecksservice/v1_store.go:94]: Intended or left-over? If the former should using logger instead?

**Review by unknwon — APPROVED** [review]
Just a minor comment.

File: cmd/enterprise-portal/internal/subscriptionlicensechecksservice/v1_store.go (hunk 3 above)
Comment (bobheadxi) [reply cmd/enterprise-portal/internal/subscriptionlicensechecksservice/v1_store.go:94]: Ah, leftover, good catch!

File: cmd/enterprise-portal/e2e/e2e_test.go (hunk 2 above)
Comment (bobheadxi) [reply cmd/enterprise-portal/e2e/e2e_test.go:87]: It's a new behaviour from the new `gprc.NewClient`:

> The target name syntax is defined in [https://github.com/grpc/grpc/blob/master/doc/naming.md](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html). e.g. to use dns resolver, a "dns:///" prefix should be applied to the target.

**Pull request merged**